### DISCUSSION TITLE
Iterating over query results using the C API

### DISCUSSION THREAD

**Discussion by joeltg** [duckdb/duckdb#3392] [Q&A]
Hi! I've been following DuckDB for a while and just for fun thought I'd try to write bindings for [Deno](https://deno.land/) using their new FFI.

I'd love to expose query results as a ES6 `Iterable` without materializing the entire result set in memory. Can it be done with the C API? It seems like the Python library has a `Cursor` class to comply with pep-0249, but is it just materializing the entire result set first and then iterating over it internally? [This discussion](https://github.com/duckdb/duckdb/discussions/2387) suggests that the JDBC driver does true iteration, but there's nothing in the C API docs that mentions cursor or iterating.

Thanks in advance!

**Comment by SartorialOffense** [comment]
I am not anything more than a DuckDb novice, but my understanding and experience is that any time you are iterating over the results by row it gets very slow. You really want to pull the data out as individual columns and access that data.  If you are allocating too much memory, you might want to try LIMIT and OFFSET to paginate your query to avoid memory problems.

**Reply by joeltg** [reply]
Ah thanks... that makes sense, maybe it's not the right thing to try to do in the first place. I was taking cues from the better-sqlite3 nodejs library which has a `Statement.iterate()` method but I don't know how well that actually performs either.
